# Ring nuclease search and characterisation pipeline (Crn4 edition)
This repo builds on top of a previous code repository that looks for known ring nucleases in prokaryotic genomes ([https://github.com/vihoikka/ring_nucleases](url)). This modified
version lacks several features from the original pipeline and is primarily used for
Crn4 analysis.

There is one main pipeline
- rn_crn4.smk: a modified ring nuclease pipeline fitted to analyse Crn4's. Looks at prokaryotic and phage genomes.

The instructions below assume basic understanding of Snakemake, python and shell/bash environment.

## Expected runtimes and results
* Running the pipeline from the beginning with 40 HPC cores takes around 48 hours.
* Due to the large number of samples which are expressed as wildcards in the Snakemake pipeline, solving the DAG (the order of rules and samples to be run) takes around 20 minutes itself. Be patient!
* The most important output file is the *mastertable_v2.tsv* in the root of the program
* Another important location is the R folder, which is used by the accompanying *rn.R* R-script

*The scripts may contain unused components*

## Requirements
* Unix based machine (tested only on Ubuntu)
* Conda
* Snakemake 7
* Python 3

*Other dependencies are installed by the pipeline when running with the --use-conda flag. Note that this will take a long time*

### Databases
#### Databases included in this repository:
* Cas10 HMM database
* Effector HMM database
* Ring nuclease HMM database
* A temperature database for prokaryotes ([TEMPURA](http://togodb.org/db/tempura))

#### Databases and software that require custom installation
* NCBI refseq genomes of bacteria and archaea (.fna, .faa and .gff files). I recommend downloading the database using ncbi-datasets (taxon 2 for bacteria; taxon 2157 for archaea):
```shell
datasets download genome taxon 2 --filename bacteria_genomes.zip --include gff3,genome,protein --dehydrated --annotated --assembly-level complete --assembly-source RefSeq
unzip bacteria_genomes.zip
datasets rehydrate --directory .
```
* Repeat the download for archaea (taxon 2157). Next, *copy* the archaeal genome folders to the bacterial folder so that they are all one big mess (*leave original archaea folders as they are*). The pipeline will use the contents of the original archaea folder and the contents of bacteria (+archaea) folder to create a file that lists the domain of each genome entry. Not too classy, but the only way I could think of how to combine all prokaryotes in the same folder without losing track of domains.
* Millard phage database. Download the latest database from Millardlab.org. If using e.g. Sept. 2024 ([link](https://millardlab.org/bacteriophage-genomics/phage-genomes-sept-2024/)), you need the files 1Sep2024_genomes.fa, 1Sep2024_vConTACT2_proteins.faa and 1Sep2024_data.tsv.
* TMHMM academic license (get from [https://services.healthtech.dtu.dk/cgi-bin/sw_request?software=tmhmm&version=2.0c&packageversion=2.0c&platform=Linux]). Once registered, you will receive a model file called TMHMM2.0.model. Place the file in the folder data/TMHMM.
* The tmhmm.py program. To install this, you need to use pip to install tmhmm.py into a Snakemake-borne Conda environment. Start the pipeline and wait until the it fails when attempting to resolve transmembrane regions in the rule cATyper_analysis. In the log file of this particular failed run (path to log file is given by Snakemake), you will find a line starting with "Current environment". This is followed by a path to the Conda environment that was active during the fail (it will be a long path, ending with something similar to ```.snakemake/conda/3d49131d367f88d1bf491d10e7655d87_```). Activate this environment manually (```conda activate <path_here>```. After activating the environment, install tmhmm.py with the command ```pip install tmhmm.py```. Now, reactivate your initial conda environment which you previously ran the pipeline with and start the pipeline again.

### Visualising results with R
Bioinformatic figure generation shown in the manuscript is provided by the R script at R/R_visualisations.R. As input for the R script, you will need several files that are generated by rn_crn4.smk:
- The Crn4 phylogenetic tree (`crn4_tree_path`) (path `base_path + "/45_effector_tree/crn4_tree.txt"`)
- Crn4 alignment data (`alignment_path`) (path `base_path + "/RN_03_aligment/crn4.afa"`)
- Crn4 genomic neighbourhood .gff files (`gff_dir`). These are found under `base_path + "/RN_06_crn4_gffs/{crn4_genome}/features_crn4_subset.gff`. Create a folder called `all` and add each .gff file there. Then point the `gff_dir` variable in the R script there.

*Note that results shown in the Crn4 distribution table (`# Create distribution table`in the R script) are hardcoded in the R script and might not represent results from your run*

### Other modifications
You will need to modify multiple paths in the snakefile to make the pipeline work on your system. You will find these lines clearly marked in the beginning of snakefile rn.smk

**Example run**
rn_crn4.smk
```snakemake --snakefile rn_crn4.smk --use-conda --cores 40 --config protein_clustering="False" getGenomesBy="local" genome_mode="all" cas10_anchor="True" --rerun-triggers mtime```

crn4_plasmid.smk

Modify the core count to match your system.

> **Note on CRISPR subtype corrections**
> After the first run, you may need to perform manual corrections of CRISPR-Cas subtypes.
> The rule mastercombiner contains a python dictionary for correcting the subtypes.
> Note that the locus id given to each locus is arbitrary and stochastic â€“ corrections therefore need to be applied every time the pipeline is run and the existing corrections in the existing codebase are very likely **not** applicable to your run. To make the manual corrections, I recommend producing a phylogenetic tree of Cas10s using the accompanying R script and manually checking if some loci seem out of place. Modify the dictionary in rule mastercombiner to include locus specific corrections and rerun starting from mastercombiner (rerunning means deleting the mastercombiner output file mastertable_v2.tsv and rerunning the pipeline. You may need to remove other downstream output files, such as the "done" file to trigger this rule again).

### The flow of the pipeline
The flow of the core pipeline in rn_crn4.smk is described in detail elsewhere (([https://github.com/vihoikka/ring_nucleases](url)). Building on top of this, the Crn4 version includes rules for generating trees and other plots in R. The relevant rules are under the header ### Crn4 specific rules ####.